fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStore.java

```diff
@@ -558,6 +565,94 @@ private Tuple recordVersionKey(@Nonnull Tuple primaryKey) {
         }
     }
 
+    @API(API.Status.EXPERIMENTAL)
+    private <M extends Message> void updateSyntheticIndexes(@Nullable FDBStoredRecord<M> oldRecord,
+                                                            @Nullable FDBStoredRecord<M> newRecord,
+                                                            @Nonnull final List<CompletableFuture<Void>> futures) {
+        final SyntheticRecordPlanner planner = new SyntheticRecordPlanner(this);
+        final int pipelineSize = getPipelineSize(PipelineOperation.IN_JOIN);
+        if (oldRecord != null && newRecord != null && oldRecord.getRecordType() == newRecord.getRecordType()) {
+            // TODO: An important optimization here is determining that no field used in the join condition or
+            //  indexed in the synthetic record is changed, in which case all this can be skipped.
+            final SyntheticRecordFromStoredRecordPlan plan = planner.fromStoredType(newRecord.getRecordType(), true);
+            if (plan == null) {
+                return;
+            }
+            final Map<RecordType, Collection<IndexMaintainer>> maintainers = getSyntheticMaintainers(plan.getSyntheticRecordTypes());
+            final Map<Tuple, FDBSyntheticRecord> oldRecords = new HashMap<>();
+            CompletableFuture<Void> future = plan.execute(this, oldRecord).forEach(syntheticRecord -> oldRecords.put(syntheticRecord.getPrimaryKey(), syntheticRecord));
+            @Nonnull final FDBStoredRecord<M> theNewRecord = newRecord; // @SpotBugsSuppressWarnings("NP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE", justification = "https://github.com/spotbugs/spotbugs/issues/552")
+            future = future.thenCompose(v -> plan.execute(this, theNewRecord).forEachAsync(syntheticRecord -> runSyntheticMaintainers(maintainers, oldRecords.remove(syntheticRecord.getPrimaryKey()), syntheticRecord), pipelineSize));
```

Comment:
I *think* this is assuming the same guarantees as `updateSecondaryIndexes`, namely that an index maintainer for a given index is not thread-safe, but that index maintainers for different indexes (non-overlapping subspaces) are, even if of the same index type.

But maybe there's an issue with a multi-type index on two of the synthetic record types for this stored record type? In which case, the maintainers don't need to be separate, but the synthetic records do?

Reply:
Can a `SyntheticRecordFromStoredRecordPlan` return multiple results on a single type index? (Like, if there's a one-to-many relationship between two record types.) My concern was that because of the parallelism of `forEachAsync`, the same maintainer might be called with different synthetic records from the same stored record concurrently.